Enable lenient parsing of Git version tags

colincornaby opened the pull request:
### Motivation:

While implementing SPM manifests in existing Git repos, I found many were using two component tags where the patch version was 0. The intention behind this change is to better support that Git tagging pattern without retraining existing teams, or causing SPM integration to be disruptive to existing tagging practices.

### Modifications:

Lenient parsing was added to swift-tools-support-core at apple/swift-tools-support-core#212. The tag based initializer for Version was altered to use this new lenient initializer.

It’s possible for two tags with the same meaning but different literal strings to be present in the same repository. This case did exist previously. Tags like v2.0.0 and 2.0.0 could conflict, and a decision about which tag to use would be required.
- Previous behavior was that a tag without a v-prefix would be preferred over one with a v-prefix. This behavior was maintained.
- A tag with more version components/more specificity is preferred. For example, 2.0.0 will be preferred over 2.0.

### Result:

Git tags with two version components will now be supported and parsed as if there was a patch version of .0 appended to the end.

tomerd commented:
@swift-ci please smoke test

abertelrud commented:
Thanks, @colincornaby!  Would you mind also adding a note to the `CHANGELOG.md` file, under the `Swift v.Next` section?  Historically we haven't been consistent about updating it when there are user-visible changes, but we are trying to get better about that.

colincornaby commented:
> Thanks, @colincornaby! Would you mind also adding a note to the `CHANGELOG.md` file, under the `Swift v.Next` section? Historically we haven't been consistent about updating it when there are user-visible changes, but we are trying to get better about that.

I added a note to the changelog and squashed it into the existing commit. Let me know if you have any feedback or edits!

WowbaggersLiquidLunch commented on CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@ Note: This is in reverse chronological order, so newer entries are added to the
 Swift v.Next
  -----------
+ * [#3649]
+     Semantic version dependencies can now be resolved against Git tag names that contain only a major and minor component.  A tag with the form `X.X` will be treated as `X.X.0`. This improves compatibility with existing repositories.
```
Just 2 nitpicks:

```suggestion
     Semantic version dependencies can now be resolved against Git tag names that contain only major and minor version identifiers.  A tag with the form `X.X` will be treated as `X.X.0`. This improves compatibility with existing repositories.
```

"identifier" is the language used in the semver specification.

```suggestion
     Semantic version dependencies can now be resolved against Git tag names that contain only a major and minor component.  A tag with the form `X.Y` will be treated as `X.Y.0`. This improves compatibility with existing repositories.
```

`X.X` might mislead users to think that the major and minor versions must be equal.

colincornaby replied:
These are good suggestions. I’ll make some edits shortly. Sorry if that’s going to reset the CI results again!

colincornaby replied:
I made the suggested edits. Thanks for the feedback!

WowbaggersLiquidLunch commented:
@swift-ci please smoke test

tomerd commented:
@swift-ci please smoke test

abertelrud commented:
> > Thanks, @colincornaby! Would you mind also adding a note to the `CHANGELOG.md` file, under the `Swift v.Next` section? Historically we haven't been consistent about updating it when there are user-visible changes, but we are trying to get better about that.
> 
> I added a note to the changelog and squashed it into the existing commit. Let me know if you have any feedback or edits!

Thanks; looks great to me!

tomerd commented:
thank you @colincornaby <3
